core/sequencer/server.go

```diff
@@ -131,55 +143,70 @@ func (s *Server) RunBatch(ctx context.Context, in *spb.RunBatchRequest) (*empty.
-	for sliceID, high := range highs {
-		sources[sliceID] = &spb.MapMetadata_SourceSlice{
-			LowestWatermark:  lastMeta.Sources[sliceID].GetHighestWatermark(),
+	count, highs, err := s.HighWatermarks(ctx, in.DomainId, startWatermarks, in.MaxBatch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "HighWatermarks(): %v", err)
+	}
+	sources := make(SourcesEntry)
+	for logID, high := range highs {
+		sources[logID] = &spb.MapMetadata_SourceSlice{
+			LowestWatermark:  lastMeta.Sources[logID].GetHighestWatermark(),
 			HighestWatermark: high,
 		}
 	}
 
-	msgs, err := s.readMessages(ctx, in.DomainId, sources)
-	if err != nil {
-		return nil, err
-	}
-	if len(msgs) < int(in.MinBatch) {
-		return &empty.Empty{}, nil
+	//
+	// Rate limit the creation of new batches.
+	///
+
+	// TODO(#1057): If time since last map revision > max timeout, run batch.
+	// TODO(#1047): If time since oldest queue item > max latency has elapsed, run batch.
+	// If count items >= min_batch, run batch.
+	if count >= in.MinBatch {
+		return s.CreateEpoch(ctx, &spb.CreateEpochRequest{
+			DomainId:    in.DomainId,
+			Revision:    int64(latestMapRoot.Revision) + 1,
+			MapMetadata: &spb.MapMetadata{Sources: sources},
+		})
 	}
 
-	return s.CreateEpoch(ctx, &spb.CreateEpochRequest{
-		DomainId:    in.DomainId,
-		Revision:    int64(latestMapRoot.Revision) + 1,
-		MapMetadata: &spb.MapMetadata{Sources: sources},
-	})
+	// TODO(#1056): If count items == max_batch, should we define the next batch immediately?
+	return &empty.Empty{}, nil
 }
 
-func (s *Server) readMessages(ctx context.Context, domainID string,
-	sources map[int64]*spb.MapMetadata_SourceSlice) ([]*ktpb.EntryUpdate, error) {
+const readBatchSize = int32(1000)
+
+// readMessages returns the full set of EntryUpdates defined by sources.
+// batchSize limits the number of messages to read from a log at one time.
+func (s *Server) readMessages(ctx context.Context, domainID string, sources SourcesEntry,
+	batchSize int32) ([]*ktpb.EntryUpdate, error) {
 	msgs := make([]*ktpb.EntryUpdate, 0)
 	for logID, source := range sources {
-		batch, err := s.logs.ReadLog(ctx, domainID, logID,
-			source.GetLowestWatermark(), source.GetHighestWatermark())
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "ReadQueue(): %v", err)
-		}
-		for _, m := range batch {
-			msgs = append(msgs, &ktpb.EntryUpdate{
-				Mutation:  m.Mutation,
-				Committed: m.ExtraData,
-			})
+		low := source.GetLowestWatermark()
+		count := readBatchSize
+		for count == readBatchSize {
+			batch, err := s.logs.ReadLog(ctx, domainID, logID, low, source.GetHighestWatermark(), batchSize)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "ReadLog(): %v", err)
+			}
+			for _, m := range batch {
+				msgs = append(msgs, &ktpb.EntryUpdate{
+					Mutation:  m.Mutation,
+					Committed: m.ExtraData,
+				})
+				if m.ID > low {
+					low = m.ID
+				}
+			}
+			count = int32(len(batch))
```

Comment:
nit: Maybe you could restructure it a bit, so that this line is merged into `for` definition?

Reply:
I did try that, but it's not very readable because `ReadLog` would need to move into the last statement in the `for` loop.  `batch` is not visible in the `for` loop scope.